File: packages/core/src/config/yaml.ts
/**
 * YAML file read/write utilities
 */
import { readFile, writeFile, mkdir } from "node:fs/promises";
import { existsSync } from "node:fs";
import { dirname } from "node:path";
import { parse, stringify } from "yaml";

/**
 * Read and parse a YAML file
 * Returns undefined if file doesn't exist
 */
export async function readYaml<T>(path: string): Promise<T | undefined> {
  try {
    if (!existsSync(path)) {
      return undefined;
    }
    const content = await readFile(path, "utf-8");
    return parse(content) as T;
  } catch (error) {
    console.error(`Failed to read YAML file: ${path}`, error);
    return undefined;
  }
}

/**
 * Write data to a YAML file
 * Creates parent directories if they don't exist
 */
export async function writeYaml<T>(path: string, data: T): Promise<void> {
  const dir = dirname(path);
  if (!existsSync(dir)) {
    await mkdir(dir, { recursive: true });
  }
  const content = stringify(data, { indent: 2 });
  await writeFile(path, content, "utf-8");
}

/**
 * Read and parse a JSON file
 * Returns undefined if file doesn't exist
 */
export async function readJson<T>(path: string): Promise<T | undefined> {
  try {
    if (!existsSync(path)) {
      return undefined;
    }
    const content = await readFile(path, "utf-8");
    return JSON.parse(content) as T;
  } catch (error) {
    console.error(`Failed to read JSON file: ${path}`, error);
    return undefined;
  }
}

/**
 * Write data to a JSON file
 * Creates parent directories if they don't exist
 */
export async function writeJson<T>(path: string, data: T): Promise<void> {
  const dir = dirname(path);
  if (!existsSync(dir)) {
    await mkdir(dir, { recursive: true });
  }
  const content = JSON.stringify(data, null, 2);
  await writeFile(path, content, "utf-8");
}

/**
 * Ensure a directory exists
 */
export async function ensureDir(path: string): Promise<void> {
  if (!existsSync(path)) {
    await mkdir(path, { recursive: true });
  }
}

/**
 * Check if a file exists
 */
export function fileExists(path: string): boolean {
  return existsSync(path);
}

File: crates/viben-core/src/config/yaml.rs
//! YAML and JSON file read/write utilities

use crate::error::Result;
use serde::{de::DeserializeOwned, Serialize};
use std::path::Path;
use tokio::fs;

/// Read and parse a YAML file
/// Returns None if file doesn't exist
pub async fn read_yaml<T: DeserializeOwned>(path: &Path) -> Result<Option<T>> {
    if !path.exists() {
        return Ok(None);
    }
    let content = fs::read_to_string(path).await?;
    let data = serde_yaml::from_str(&content)?;
    Ok(Some(data))
}

/// Write data to a YAML file
/// Creates parent directories if they don't exist
pub async fn write_yaml<T: Serialize>(path: &Path, data: &T) -> Result<()> {
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent).await?;
    }
    let content = serde_yaml::to_string(data)?;
    fs::write(path, content).await?;
    Ok(())
}

/// Read and parse a JSON file
/// Returns None if file doesn't exist
pub async fn read_json<T: DeserializeOwned>(path: &Path) -> Result<Option<T>> {
    if !path.exists() {
        return Ok(None);
    }
    let content = fs::read_to_string(path).await?;
    let data = serde_json::from_str(&content)?;
    Ok(Some(data))
}

/// Write data to a JSON file
/// Creates parent directories if they don't exist
pub async fn write_json<T: Serialize>(path: &Path, data: &T) -> Result<()> {
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent).await?;
    }
    let content = serde_json::to_string_pretty(data)?;
    fs::write(path, content).await?;
    Ok(())
}

/// Ensure a directory exists
pub async fn ensure_dir(path: &Path) -> Result<()> {
    if !path.exists() {
        fs::create_dir_all(path).await?;
    }
    Ok(())
}

/// Check if a file exists
pub fn file_exists(path: &Path) -> bool {
    path.exists()
}

/// Check if a directory exists
pub fn dir_exists(path: &Path) -> bool {
    path.is_dir()
}
